### ISSUE TITLE
Propagating long times does not converge

### ISSUE THREAD

**Issue by Juanlu001** [poliastro/poliastro#265]
🐞 **Problem**

Propagating long times fails in different ways:

```python
In [5]: iss.propagate(100 * u.year)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-27-503eb53e62e7> in <module>()
----> 1 iss.propagate(100 * u.year)

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in propagate(self, epoch_or_duration, rtol)
    245             time_of_flight = time.TimeDelta(epoch_or_duration)
    246 
--> 247         return propagate(self, time_of_flight, rtol=rtol)
    248 
    249     def sample(self, *, num_points=None, time_values=None):

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in propagate(orbit, time_of_flight, method, rtol, **kwargs)
    156                   time_of_flight.to(u.s).value,
    157                   rtol=rtol,
--> 158                   **kwargs)
    159     return orbit.from_vectors(orbit.attractor, r * u.km, v * u.km / u.s, orbit.epoch + time_of_flight)
    160 

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in kepler(k, r0, v0, tof, rtol, numiter)
    139     f, g, fdot, gdot = _kepler(k, r0, v0, tof, numiter, rtol)
    140 
--> 141     assert np.abs(f * gdot - fdot * g - 1) < 1e-5  # Fixed tolerance
    142 
    143     # Return position and velocity vectors

AssertionError: 

In [6]: halleys[0].propagate(73 * u.year)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-26-9dcacad8e904> in <module>()
----> 1 halleys[0].propagate(73 * u.year)

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in propagate(self, epoch_or_duration, rtol)
    245             time_of_flight = time.TimeDelta(epoch_or_duration)
    246 
--> 247         return propagate(self, time_of_flight, rtol=rtol)
    248 
    249     def sample(self, *, num_points=None, time_values=None):

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in propagate(orbit, time_of_flight, method, rtol, **kwargs)
    156                   time_of_flight.to(u.s).value,
    157                   rtol=rtol,
--> 158                   **kwargs)
    159     return orbit.from_vectors(orbit.attractor, r * u.km, v * u.km / u.s, orbit.epoch + time_of_flight)
    160 

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in kepler(k, r0, v0, tof, rtol, numiter)
    137     """
    138     # Compute Lagrange coefficients
--> 139     f, g, fdot, gdot = _kepler(k, r0, v0, tof, numiter, rtol)
    140 
    141     assert np.abs(f * gdot - fdot * g - 1) < 1e-5  # Fixed tolerance

RuntimeError: Maximum number of iterations reached
```

🎯 **Goal**

One thing is that the long term errors of propagation are big. But nonetheless, the propagator should always give an answer, instead of an internal error (at least for a full period, see the failure of the Halley's comet). I wonder if this is possible at all?

💡 **Possible solutions**

The propagator of poliastro uses the universal variables approach, which is popular but has several limitations. Perhaps for long propagation times it should be replaced with a Newton-Raphson iteration and a good starting point as suggested by http://dx.doi.org/10.1007/s10569-013-9476-9.

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

**Comment by Juanlu001** [comment]
Related: #253

**Comment by Juanlu001** [comment]
This would be a great moment to validate against SPICE or GMAT. The problems of validating against such complex software is that some parameters are difficult to control, but there is no alternative since we are already stepping outside of the "comfort zone" of easy problems that can be validated using textbook assignments, or even research papers (which don't always provide enough information themselves, see https://github.com/Juanlu001/pfc-uc3m).

**Comment by nikita-astronaut** [comment]
I will try to investigate this issue too. I'm specialized in numerical analysis, maybe I fill figure it out.

**Comment by nikita-astronaut** [comment]
@Juanlu001 , I have a question on how to implement this.
Given Cartesian elements (\vec{r}, \vec{v}) at the initial moment, I first calculate all the orbit parameters. Then, using (4) from the paper I get M_E at time (t + Delta_t) and iteratively solve E - e \sin E = M_E. After that, I have (E, M_E) and all the orbit parameters. And I use some formulae (that of course are written somewhere) to get back to (\vec{r}, \vec{v}). Is this logic OK?

**Comment by nikita-astronaut** [comment]
Actually, in (4) there is t_p, the time at which the body passes nearest point to attractor. Is there an efficient or already implemented in this package way to find it?

**Comment by Juanlu001** [comment]
The conversion between cartesian elements and Keplerian elements is in the coe2rv/rv2coe functions. You could also create an `Orbit` from one set of elements and then retrieve the other.

Regarding the time of pericenter passage, it's not implemented but it would be a good addition to the `Orbit` object. Perhaps you could start there.

**Comment by nikita-astronaut** [comment]
Well, I realized there is no need to find the pericenter passage time. The rv2coe gives current true anomaly $\nu(t_0)$, which can be recalculated into current mean anomaly $M(t_0)$. Then, I recall that mean anomaly grows linearly with time, which gives me mean anomaly at the time $t_0 + \Deltat$: $M(t_0 + \Delta t) = M(t_0) + \sqrt{\mu (1 - e)^3 / q^3} \Delta t$. As I have $M(t_0 + \Delta t)$, I iteratively solve $E - e \sin E = M$, get $\nu$ (this is actually already implemented within M_to_nu) and thus new r, v. 

I think I'll do it this way.

**Comment by Juanlu001** [comment]
This was addressed in #322, can someone take a screenshot of a dense sampling of the Halley orbits as seen here http://docs.poliastro.space/en/latest/examples/Using%20NEOS%20package.html?highlight=halley#DASTCOM5-module to check that this actually works?

**Comment by Juanlu001** [comment]
I confirm that the original `iss.propagate(100 * u.year)` example works. I'm adding some regression tests in a pull request.

**Comment by Juanlu001** [comment]
Notice that the original convergence issues of `kepler` were addressed in #362.
